homeassistant/components/binary_sensor/ihc.py

```diff
@@ -0,0 +1,183 @@
+    ihcplatform = get_ihc_platform(hass)
+    devices = []
+    if config.get(CONF_AUTOSETUP):
+        auto_setup(ihcplatform, devices)
+
+    binarysensors = config.get(CONF_BINARY_SENSORS)
+    if binarysensors:
+        _LOGGER.info("Adding IHC Binary Sensors")
+        for binarysensor in binarysensors:
+            ihcid = binarysensor[CONF_ID]
+            name = (binarysensor[CONF_NAME]
+                    if CONF_NAME in binarysensor
+                    else "ihc_" + str(ihcid))
+            sensortype = (binarysensor[CONF_TYPE]
+                          if CONF_TYPE in binarysensor
+                          else None)
+            inverting = (binarysensor[CONF_INVERTING]
+                         if CONF_INVERTING in binarysensor
+                         else False)
+            add_sensor(devices, ihcplatform.ihc, int(ihcid), name, sensortype,
+                       True, inverting)
+
+    add_devices(devices)
+    # Start notification after devices has been added
+    for device in devices:
+        device.ihc.add_notify_event(device.get_ihcid(),
+                                    device.on_ihc_change, True)
+
+
+def auto_setup(ihcplatform, devices):
+    """Auto setup ihc binary sensors from ihc project."""
+    _LOGGER.info("Auto setup - IHC Binary sensors")
+
+    def setup_product(ihcid, name, product, productcfg):
+        """Product setup callback."""
+        add_sensor_from_node(devices, ihcplatform.ihc, ihcid, name,
+                             product, productcfg['type'],
+                             productcfg['inverting'])
+    ihcplatform.autosetup(PRODUCTAUTOSETUP, setup_product)
+
+
+class IHCBinarySensor(IHCDevice, BinarySensorDevice):
+    """IHC Binary Sensor."""
+
+    def __init__(self, ihccontroller, name, ihcid, sensortype: str,
+                 inverting: bool, ihcname: str, ihcnote: str,
+                 ihcposition: str):
+        """Initialize the IHC binary sensor."""
+        IHCDevice.__init__(self, ihccontroller, name, ihcid, ihcname,
+                           ihcnote, ihcposition)
+        self._state = STATE_UNKNOWN
+        self._sensor_type = sensortype
+        self.inverting = inverting
+
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        return False
+
+    @property
+    def device_class(self):
+        """Return the class of this sensor."""
+        return self._sensor_type
+
+    @property
+    def is_on(self):
+        """Return true if the binary sensor is on/open."""
+        return self._state
+
+    def update(self):
+        """We do not nothing."""
+        pass
+
+    def on_ihc_change(self, ihcid, value):
+        """Callback when ihc resource changes."""
+        if self.inverting:
+            self._state = not value
+        else:
+            self._state = value
+        self.schedule_update_ha_state()
+
+
+def add_sensor_from_node(devices, ihccontroller, ihcid: int, name: str,
```

Comment:
I think you should be able to move this function to the ihc component and make it reusable by all ihc platforms.

Reply:
Done (Product auto setup completely changed - see other comment)